src/prommis/nanofiltration/diafiltration_two_salt.py

```diff
@@ -1096,27 +1096,65 @@ def add_scaling_factors(self):
 
     def add_ports(self):
         self.feed_inlet = Port(doc="Feed Inlet Port")
-        self.feed_flow_volume_ref = Reference(self.feed_flow_volume)
-        self.feed_inlet.add(self.feed_flow_volume_ref, "flow_vol")
-        self.feed_conc_mass_comp_ref = Reference(self.feed_conc_mass_comp)
-        self.feed_inlet.add(self.feed_conc_mass_comp_ref, "conc_mass_comp")
+        self._feed_flow_volume_ref = Reference(self.feed_flow_volume)
+        self.feed_inlet.add(self._feed_flow_volume_ref, "flow_vol")
+        self._feed_conc_mass_comp_ref = Reference(self.feed_conc_mass_comp)
+        self.feed_inlet.add(self._feed_conc_mass_comp_ref, "conc_mass_comp")
 
         self.diafiltrate_inlet = Port(doc="Diafiltrate Inlet Port")
-        self.diafiltrate_flow_volume_ref = Reference(self.diafiltrate_flow_volume)
-        self.diafiltrate_inlet.add(self.diafiltrate_flow_volume_ref, "flow_vol")
-        self.diafiltrate_conc_mass_comp_ref = Reference(self.diafiltrate_conc_mass_comp)
+        self._diafiltrate_flow_volume_ref = Reference(self.diafiltrate_flow_volume)
+        self.diafiltrate_inlet.add(self._diafiltrate_flow_volume_ref, "flow_vol")
+        self._diafiltrate_conc_mass_comp_ref = Reference(
+            self.diafiltrate_conc_mass_comp
+        )
         self.diafiltrate_inlet.add(
-            self.diafiltrate_conc_mass_comp_ref, "conc_mass_comp"
+            self._diafiltrate_conc_mass_comp_ref, "conc_mass_comp"
         )
 
         self.retentate_outlet = Port(doc="Retentate Outlet Port")
-        self.retentate_flow_volume_ref = Reference(self.retentate_flow_volume)
-        self.retentate_outlet.add(self.retentate_flow_volume_ref, "flow_vol")
-        self.retentate_conc_mass_comp_ref = Reference(self.retentate_conc_mass_comp)
-        self.retentate_outlet.add(self.retentate_conc_mass_comp_ref, "conc_mass_comp")
+        self._retentate_flow_volume_ref = Reference(
+            self.retentate_flow_volume[0, value(self.membrane_width)]
+        )
+        self.retentate_outlet.add(self._retentate_flow_volume_ref, "flow_vol")
+        self._retentate_conc_mass_lithium_ref = Reference(
+            self.retentate_conc_mass_comp[0, value(self.membrane_width), "Li"]
+        )
+        self.retentate_outlet.add(
+            self._retentate_conc_mass_lithium_ref, "conc_mass_lithium"
+        )
+        self._retentate_conc_mass_cobalt_ref = Reference(
+            self.retentate_conc_mass_comp[0, value(self.membrane_width), "Co"]
+        )
+        self.retentate_outlet.add(
+            self._retentate_conc_mass_cobalt_ref, "conc_mass_cobalt"
+        )
+        self._retentate_conc_mass_chlorine_ref = Reference(
+            self.retentate_conc_mass_comp[0, value(self.membrane_width), "Cl"]
+        )
+        self.retentate_outlet.add(
+            self._retentate_conc_mass_chlorine_ref, "conc_mass_chlorine"
+        )
```

Comment:
You didn't need to break `conc_mass_comp` into element-by-element variables. You can create `References` to slices. For example, you could create `self._retentate_conc_mass_comp_ref = Reference(self.retentate_conc_mass_comp[:, self.x_bar.last(), :])` to collapse the length dimension while retaining the time and component dimensions.

You should make sure to use `x_bar.last()` , because that's always going to give the right answer. You're working with a reduced length domain, and `membrane_width` is a `Var`, so `value` may not even return a member of `x_bar`. (I think that's a moot point in this case because `membrane_width` is initialized to 1 and probably can't be changed by the user before these ports are created, but it's still bad form).